2020 Alabama State House/Senate Districts

Kara77779 opened the pull request:
# 2020 Alabama State House/Senate Districts

## Redistricting requirements
In Alabama, we consult [NCSL Redistricting Law 2020](https://documents.ncsl.org/wwwncsl/Redistricting-Census/Redistricting-Law-2020_NCSL%20FINAL.pdf) and impose the following constraints. In our simulations, legislative districts must:

1. Be contiguous [NCSL, 184]
2. Have equal populations [NCSL, 23]
3. Be geographically compact [NCSL, 184]
4. Preserve political subdivisions and communities of interest [NCSL, 184]

### Algorithmic Constraints
We enforce a maximum population deviation of 5.0%.

## Data Sources
Data for Alabama comes from the ALARM Project's [2020 Redistricting Data Files](https://alarm-redist.github.io/posts/2021-08-10-census-2020/).

## Pre-processing Notes
No manual pre-processing decisions were necessary.

## Simulation Notes
We sample 25,000 districting plans for Alabama's lower house across 5 independent runs of the SMC algorithm.
We then thinned the number of samples to 10,000.
No special techniques were needed to produce the sample.

We sample 10,000 districting plans for Alabama's upper house across 5 independent runs of the SMC algorithm.
No special techniques were needed to produce the sample.

## Validation
## SHD

**<img width="3000" height="5400" alt="validation_20251221_1853" src="https://github.com/user-attachments/assets/7b408f12-72a1-472d-a2f2-92437c90b62f" />**


```
SMC with Merge-Split MCMC Steps: 10,000 sampled plans of 105 districts on 1,837 units
Plans sampled on Linking Edge Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:                       
• `weight_type` = optimal             
• `seq_alpha` = 1                     
• `pop_temper` = 0                    
Mergesplit Parameters:                
• `total_ms_steps` = 104              
• `mh_accept_per_smc` = 110           
• `pair_rule` = uniform               
Plan diversity 80% range: 0.71 to 0.79
85 rhat values are `NA`or AL_shd_2020

Largest R-hat values for ordered summary statistics:
        adv_16         adv_18         adv_20         arv_16         arv_18         arv_20 atg_18_dem_sie 
         1.016          1.027          1.016          1.017          1.015          1.016          1.029 
atg_18_rep_mar      comp_edge    comp_polsby  county_splits          e_dem          e_dvs           egap 
         1.012          1.016          1.011          1.021          1.005          1.022          1.005 
gov_18_dem_mad gov_18_rep_ive    muni_splits        ndshare            ndv            nrv          pbias 
         1.019          1.017          1.011          1.022          1.026          1.015          1.017 
      plan_dev       pop_aian      pop_asian      pop_black       pop_hisp       pop_nhpi      pop_other 
         1.001          1.038          1.023          1.023          1.031          1.037          1.027 
   pop_overlap        pop_two      pop_white         pr_dem pre_16_dem_cli pre_16_rep_tru pre_20_dem_bid 
         1.015          1.020          1.011          1.007          1.017          1.015          1.018 
pre_20_rep_tru sos_18_dem_mil sos_18_rep_mer      total_vap uss_16_dem_cru uss_16_rep_she uss_20_dem_jon 
         1.015          1.025          1.016          1.017          1.015          1.024          1.014 
uss_20_rep_tub       vap_aian      vap_asian      vap_black       vap_hisp       vap_nhpi      vap_other 
         1.017          1.017          1.030          1.023          1.028          1.014          1.018 
       vap_two      vap_white 
         1.017          1.011 
• R-hat ≤ 1.05: 5270                  
• 1.05 < R-hat ≤ 1.1: 0               
• R-hat > 1.1: 0                      
• Total R-hats: 5270                  
Sampling diagnostics for SMC run 1 of 5 (10,000 samples)
          Eff. samples (%) Acc. rate Log wgt. sd  Max. unique 
Split 1      2,321 (46.4%)    100.0%        1.36 3,146 (100%) 
Split 2      2,501 (50.0%)     99.0%        1.30 2,474 ( 78%) 
Split 3      2,744 (54.9%)     98.0%        1.19 2,470 ( 78%) 
Split 4      2,984 (59.7%)     97.4%        1.09 2,568 ( 81%) 
Split 5      3,051 (61.0%)     96.1%        1.01 2,629 ( 83%) 
Split 6      3,114 (62.3%)     94.9%        0.97 2,675 ( 85%) 
Split 7      3,282 (65.6%)     93.6%        0.91 2,675 ( 85%) 
Split 8      3,339 (66.8%)     92.4%        0.88 2,740 ( 87%) 
Split 9      3,460 (69.2%)     91.7%        0.81 2,768 ( 88%) 
Split 10     3,509 (70.2%)     90.1%        0.78 2,824 ( 89%) 
Split 11     3,577 (71.5%)     89.3%        0.75 2,830 ( 90%) 
Split 12     3,710 (74.2%)     88.1%        0.70 2,833 ( 90%) 
Split 13     3,745 (74.9%)     87.1%        0.68 2,867 ( 91%) 
Split 14     3,830 (76.6%)     85.7%        0.66 2,889 ( 91%) 
Split 15     3,909 (78.2%)     83.5%        0.61 2,949 ( 93%) 
Split 16     3,958 (79.2%)     83.2%        0.60 2,926 ( 93%) 
Split 17     3,977 (79.5%)     81.7%        0.58 2,925 ( 93%) 
Split 18     4,050 (81.0%)     80.9%        0.57 2,957 ( 94%) 
Split 19     4,086 (81.7%)     79.1%        0.55 2,948 ( 93%) 
Split 20     4,126 (82.5%)     79.9%        0.52 3,005 ( 95%) 
Split 21     4,185 (83.7%)     76.4%        0.51 2,971 ( 94%) 
Split 22     4,182 (83.6%)     75.8%        0.50 2,992 ( 95%) 
Split 23     4,240 (84.8%)     75.0%        0.47 3,065 ( 97%) 
Split 24     4,310 (86.2%)     74.0%        0.46 3,018 ( 95%) 
Split 25     4,299 (86.0%)     72.6%        0.45 3,046 ( 96%) 
Split 26     4,313 (86.3%)     71.8%        0.45 2,982 ( 94%) 
Split 27     4,343 (86.9%)     70.5%        0.43 2,996 ( 95%) 
Split 28     4,376 (87.5%)     68.7%        0.42 3,039 ( 96%) 
Split 29     4,374 (87.5%)     69.2%        0.42 2,998 ( 95%) 
Split 30     4,435 (88.7%)     67.5%        0.40 3,040 ( 96%) 
Split 31     4,431 (88.6%)     65.5%        0.40 3,071 ( 97%) 
Split 32     4,480 (89.6%)     65.6%        0.38 3,060 ( 97%) 
Split 33     4,457 (89.1%)     64.9%        0.38 3,066 ( 97%) 
Split 34     4,504 (90.1%)     63.2%        0.37 3,033 ( 96%) 
Split 35     4,498 (90.0%)     61.8%        0.36 3,090 ( 98%) 
Split 36     4,526 (90.5%)     60.9%        0.36 3,028 ( 96%) 
Split 37     4,559 (91.2%)     61.0%        0.35 3,048 ( 96%) 
Split 38     4,560 (91.2%)     59.8%        0.34 3,048 ( 96%) 
Split 39     4,583 (91.7%)     58.2%        0.33 3,090 ( 98%) 
Split 40     4,575 (91.5%)     57.2%        0.34 3,068 ( 97%) 
Split 41     4,581 (91.6%)     57.1%        0.33 3,099 ( 98%) 
Split 42     4,599 (92.0%)     56.1%        0.32 3,080 ( 97%) 
Split 43     4,595 (91.9%)     54.5%        0.33 3,073 ( 97%) 
Split 44     4,616 (92.3%)     53.9%        0.31 3,091 ( 98%) 
Split 45     4,634 (92.7%)     53.4%        0.30 3,060 ( 97%) 
Split 46     4,654 (93.1%)     52.3%        0.30 3,088 ( 98%) 
Split 47     4,649 (93.0%)     51.8%        0.30 3,051 ( 97%) 
Split 48     4,653 (93.1%)     50.5%        0.29 3,102 ( 98%) 
Split 49     4,676 (93.5%)     50.0%        0.28 3,091 ( 98%) 
Split 50     4,672 (93.4%)     48.8%        0.29 3,094 ( 98%) 
Split 51     4,690 (93.8%)     48.9%        0.27 3,050 ( 97%) 
Split 52     4,691 (93.8%)     46.9%        0.28 3,076 ( 97%) 
Split 53     4,704 (94.1%)     47.0%        0.27 3,058 ( 97%) 
Split 54     4,708 (94.2%)     45.1%        0.27 3,069 ( 97%) 
Split 55     4,720 (94.4%)     45.3%        0.26 3,093 ( 98%) 
Split 56     4,726 (94.5%)     44.3%        0.26 3,117 ( 99%) 
Split 57     4,729 (94.6%)     43.5%        0.26 3,083 ( 98%) 
Split 58     4,736 (94.7%)     42.9%        0.25 3,077 ( 97%) 
Split 59     4,754 (95.1%)     42.3%        0.24 3,088 ( 98%) 
Split 60     4,752 (95.0%)     40.5%        0.24 3,087 ( 98%) 
Split 61     4,765 (95.3%)     40.8%        0.24 3,088 ( 98%) 
Split 62     4,762 (95.2%)     40.7%        0.24 3,099 ( 98%) 
Split 63     4,765 (95.3%)     39.6%        0.24 3,130 ( 99%) 
Split 64     4,762 (95.2%)     39.7%        0.23 3,120 ( 99%) 
Split 65     4,775 (95.5%)     38.8%        0.23 3,101 ( 98%) 
Split 66     4,774 (95.5%)     38.1%        0.23 3,107 ( 98%) 
Split 67     4,793 (95.9%)     37.2%        0.22 3,110 ( 98%) 
Split 68     4,794 (95.9%)     36.7%        0.22 3,098 ( 98%) 
Split 69     4,800 (96.0%)     36.2%        0.22 3,086 ( 98%) 
Split 70     4,803 (96.1%)     35.8%        0.21 3,112 ( 98%) 
Split 71     4,809 (96.2%)     35.2%        0.21 3,062 ( 97%) 
Split 72     4,814 (96.3%)     34.6%        0.21 3,061 ( 97%) 
Split 73     4,816 (96.3%)     34.1%        0.21 3,101 ( 98%) 
Split 74     4,826 (96.5%)     33.6%        0.20 3,116 ( 99%) 
Split 75     4,824 (96.5%)     32.9%        0.20 3,043 ( 96%) 
Split 76     4,829 (96.6%)     32.9%        0.20 3,102 ( 98%) 
Split 77     4,833 (96.7%)     31.3%        0.20 3,080 ( 97%) 
Split 78     4,837 (96.7%)     31.2%        0.19 3,080 ( 97%) 
Split 79     4,836 (96.7%)     31.3%        0.19 3,106 ( 98%) 
Split 80     4,845 (96.9%)     30.0%        0.19 3,078 ( 97%) 
Split 81     4,841 (96.8%)     29.6%        0.19 3,099 ( 98%) 
Split 82     4,843 (96.9%)     29.5%        0.19 3,094 ( 98%) 
Split 83     4,846 (96.9%)     29.0%        0.19 3,091 ( 98%) 
Split 84     4,846 (96.9%)     28.0%        0.19 3,080 ( 97%) 
Split 85     4,851 (97.0%)     27.8%        0.18 3,121 ( 99%) 
Split 86     4,860 (97.2%)     27.4%        0.18 3,081 ( 97%) 
Split 87     4,865 (97.3%)     26.6%        0.17 3,130 ( 99%) 
Split 88     4,860 (97.2%)     26.5%        0.18 3,090 ( 98%) 
Split 89     4,865 (97.3%)     26.1%        0.17 3,077 ( 97%) 
Split 90     4,873 (97.5%)     25.3%        0.17 3,086 ( 98%) 
Split 91     4,878 (97.6%)     25.1%        0.17 3,076 ( 97%) 
Split 92     4,873 (97.5%)     24.2%        0.17 3,087 ( 98%) 
Split 93     4,876 (97.5%)     23.8%        0.17 3,064 ( 97%) 
Split 94     4,876 (97.5%)     24.1%        0.17 3,092 ( 98%) 
Split 95     4,881 (97.6%)     23.5%        0.16 3,041 ( 96%) 
Split 96     4,880 (97.6%)     23.2%        0.16 3,070 ( 97%) 
Split 97     4,886 (97.7%)     22.7%        0.16 3,043 ( 96%) 
Split 98     4,882 (97.6%)     22.3%        0.16 3,001 ( 95%) 
Split 99     4,886 (97.7%)     21.8%        0.16 2,989 ( 95%) 
Split 100    4,891 (97.8%)     21.8%        0.16 2,964 ( 94%) 
Split 101    4,896 (97.9%)     21.3%        0.15 2,841 ( 90%) 
Split 102    4,901 (98.0%)     20.6%        0.15 2,824 ( 89%) 
Split 103    4,911 (98.2%)     21.3%        0.14 2,749 ( 87%) 
Split 104    4,913 (98.3%)     20.3%        0.14 2,465 ( 78%) 
Resample     4,913 (98.3%)       NA%          NA 4,709 (149%) 

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 5 (10,000 samples)
            Acc. rate MS Steps per Plan
MS Step 1       48.3%               110
MS Step 2       48.2%               330
MS Step 3       48.3%               330
MS Step 4       48.4%               330
MS Step 5       48.4%               330
MS Step 6       48.4%               330
MS Step 7       48.3%               330
MS Step 8       48.4%               330
MS Step 9       48.2%               330
MS Step 10      48.1%               330
MS Step 11      48.0%               330
MS Step 12      47.8%               330
MS Step 13      47.7%               330
MS Step 14      47.6%               330
MS Step 15      47.4%               330
MS Step 16      47.2%               330
MS Step 17      47.0%               330
MS Step 18      47.1%               330
MS Step 19      46.9%               330
MS Step 20      46.6%               330
MS Step 21      46.3%               330
MS Step 22      46.2%               330
MS Step 23      46.1%               330
MS Step 24      45.9%               330
MS Step 25      45.7%               330
MS Step 26      45.5%               330
MS Step 27      45.3%               330
MS Step 28      45.1%               330
MS Step 29      44.9%               330
MS Step 30      44.6%               330
MS Step 31      44.6%               330
MS Step 32      44.2%               330
MS Step 33      44.1%               330
MS Step 34      44.0%               330
MS Step 35      43.7%               330
MS Step 36      43.6%               330
MS Step 37      43.3%               330
MS Step 38      43.2%               330
MS Step 39      42.9%               330
MS Step 40      42.6%               330
MS Step 41      42.6%               330
MS Step 42      42.4%               330
MS Step 43      42.2%               330
MS Step 44      41.9%               330
MS Step 45      41.9%               330
MS Step 46      41.5%               330
MS Step 47      41.4%               330
MS Step 48      41.2%               330
MS Step 49      40.9%               330
MS Step 50      40.8%               330
MS Step 51      40.6%               330
MS Step 52      40.4%               330
MS Step 53      40.3%               330
MS Step 54      40.0%               330
MS Step 55      39.8%               330
MS Step 56      39.6%               330
MS Step 57      39.4%               330
MS Step 58      39.2%               330
MS Step 59      39.1%               330
MS Step 60      38.9%               330
MS Step 61      38.7%               330
MS Step 62      38.5%               330
MS Step 63      38.4%               330
MS Step 64      38.1%               330
MS Step 65      37.9%               330
MS Step 66      37.8%               330
MS Step 67      37.7%               330
MS Step 68      37.5%               330
MS Step 69      37.3%               330
MS Step 70      37.1%               330
MS Step 71      36.9%               330
MS Step 72      36.8%               330
MS Step 73      36.8%               330
MS Step 74      36.5%               330
MS Step 75      36.3%               330
MS Step 76      36.1%               330
MS Step 77      36.0%               330
MS Step 78      35.8%               330
MS Step 79      35.7%               330
MS Step 80      35.6%               330
MS Step 81      35.3%               330
MS Step 82      35.1%               330
MS Step 83      35.0%               330
MS Step 84      34.9%               330
MS Step 85      34.8%               330
MS Step 86      34.6%               330
MS Step 87      34.4%               330
MS Step 88      34.3%               330
MS Step 89      34.2%               330
MS Step 90      34.0%               330
MS Step 91      33.9%               330
MS Step 92      33.7%               330
MS Step 93      33.6%               330
MS Step 94      33.6%               330
MS Step 95      33.3%               330
MS Step 96      33.2%               440
MS Step 97      33.1%               440
MS Step 98      33.0%               440
MS Step 99      32.9%               440
MS Step 100     32.7%               440
MS Step 101     32.5%               440
MS Step 102     32.3%               440
MS Step 103     32.2%               440
MS Step 104     32.0%               440

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of the log
weights (more than 3 or so), and low numbers of unique plans. R-hat values for summary statistics should be
between 1 and 1.05.
```

## SSD

<img width="3000" height="5400" alt="validation_20251222_1533" src="https://github.com/user-attachments/assets/24b0a6aa-de0e-4efa-9ad2-d28fe36ace60" />

```
SMC with Merge-Split MCMC Steps: 10,000 sampled plans of 35 districts on 1,837 units
Plans sampled on Linking Edge Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:
• `weight_type` = optimal
• `seq_alpha` = 1
• `pop_temper` = 0
Mergesplit Parameters:
• `total_ms_steps` = 34
• `mh_accept_per_smc` = 12
• `pair_rule` = uniform
Plan diversity 80% range: 0.73 to 0.85
23 rhat values are `NA`
Largest R-hat values for ordered summary statistics:
             adv_16              adv_18              adv_20              arv_16              arv_18              arv_20 
              1.017               1.022               1.021               1.014               1.014               1.013 
     atg_18_dem_sie      atg_18_rep_mar     comp_bbox_reock           comp_edge         comp_polsby       county_splits 
              1.023               1.014               1.020               1.000               1.015               1.012 
              e_dem               e_dvs                egap      gov_18_dem_mad      gov_18_rep_ive         muni_splits 
              1.010               1.040               1.013               1.018               1.014               1.006 
            ndshare                 ndv                 nrv               pbias            plan_dev            pop_aian 
              1.039               1.029               1.014               1.009               1.000               1.022 
          pop_asian           pop_black            pop_hisp            pop_nhpi           pop_other         pop_overlap 
              1.017               1.024               1.021               1.020               1.020               1.018 
            pop_two           pop_white              pr_dem      pre_16_dem_cli      pre_16_rep_tru      pre_20_dem_bid 
              1.020               1.013               1.012               1.034               1.014               1.023 
     pre_20_rep_tru      sos_18_dem_mil      sos_18_rep_mer total_county_splits   total_muni_splits           total_vap 
              1.013               1.031               1.014               1.016               1.007               1.004 
     uss_16_dem_cru      uss_16_rep_she      uss_20_dem_jon      uss_20_rep_tub            vap_aian           vap_asian 
              1.017               1.014               1.020               1.013               1.015               1.020 
          vap_black            vap_hisp            vap_nhpi           vap_other             vap_two           vap_white 
              1.020               1.018               1.017               1.017               1.015               1.010 
• R-hat ≤ 1.05: 1867
• 1.05 < R-hat ≤ 1.1: 0
• R-hat > 1.1: 0
• Total R-hats: 1867
Sampling diagnostics for SMC run 1 of 5 (10,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique 
Split 1       936 (46.8%)    100.0%        1.25 1,271 (101%) 
Split 2       985 (49.3%)     98.3%        1.22   960 ( 76%) 
Split 3     1,116 (55.8%)     95.1%        1.11   996 ( 79%) 
Split 4     1,206 (60.3%)     92.2%        0.99 1,027 ( 81%) 
Split 5     1,244 (62.2%)     88.4%        0.94 1,078 ( 85%) 
Split 6     1,303 (65.2%)     86.6%        0.88 1,057 ( 84%) 
Split 7     1,366 (68.3%)     83.4%        0.80 1,107 ( 88%) 
Split 8     1,384 (69.2%)     79.7%        0.78 1,093 ( 86%) 
Split 9     1,412 (70.6%)     76.0%        0.76 1,111 ( 88%) 
Split 10    1,478 (73.9%)     73.7%        0.68 1,118 ( 88%) 
Split 11    1,508 (75.4%)     72.3%        0.65 1,136 ( 90%) 
Split 12    1,545 (77.2%)     67.8%        0.62 1,142 ( 90%) 
Split 13    1,595 (79.8%)     65.3%        0.57 1,165 ( 92%) 
Split 14    1,616 (80.8%)     63.6%        0.54 1,180 ( 93%) 
Split 15    1,617 (80.8%)     60.2%        0.55 1,178 ( 93%) 
Split 16    1,634 (81.7%)     58.1%        0.53 1,184 ( 94%) 
Split 17    1,679 (83.9%)     56.2%        0.49 1,170 ( 93%) 
Split 18    1,690 (84.5%)     54.0%        0.48 1,188 ( 94%) 
Split 19    1,695 (84.8%)     51.7%        0.45 1,198 ( 95%) 
Split 20    1,726 (86.3%)     50.5%        0.44 1,212 ( 96%) 
Split 21    1,730 (86.5%)     47.3%        0.43 1,186 ( 94%) 
Split 22    1,742 (87.1%)     47.0%        0.41 1,215 ( 96%) 
Split 23    1,754 (87.7%)     44.1%        0.40 1,180 ( 93%) 
Split 24    1,761 (88.0%)     43.1%        0.40 1,173 ( 93%) 
Split 25    1,777 (88.8%)     40.7%        0.38 1,208 ( 96%) 
Split 26    1,796 (89.8%)     39.0%        0.36 1,189 ( 94%) 
Split 27    1,820 (91.0%)     38.3%        0.34 1,183 ( 94%) 
Split 28    1,814 (90.7%)     36.6%        0.34 1,203 ( 95%) 
Split 29    1,831 (91.5%)     34.2%        0.33 1,197 ( 95%) 
Split 30    1,850 (92.5%)     33.8%        0.31 1,218 ( 96%) 
Split 31    1,862 (93.1%)     33.3%        0.30 1,198 ( 95%) 
Split 32    1,871 (93.5%)     31.1%        0.29 1,191 ( 94%) 
Split 33    1,892 (94.6%)     30.1%        0.27 1,168 ( 92%) 
Split 34    1,909 (95.5%)     29.7%        0.24 1,054 ( 83%) 
Resample    1,909 (95.5%)       NA%          NA 1,831 (145%) 

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 5 (10,000 samples)
           Acc. rate MS Steps per Plan
MS Step 1      54.7%                12
MS Step 2      52.1%                24
MS Step 3      50.4%                24
MS Step 4      50.0%                24
MS Step 5      50.0%                24
MS Step 6      48.5%                24
MS Step 7      47.3%                36
MS Step 8      47.1%                36
MS Step 9      45.8%                36
MS Step 10     44.9%                36
MS Step 11     44.3%                36
MS Step 12     43.4%                36
MS Step 13     43.0%                36
MS Step 14     41.7%                36
MS Step 15     41.0%                36
MS Step 16     40.5%                36
MS Step 17     39.7%                36
MS Step 18     39.2%                36
MS Step 19     37.9%                36
MS Step 20     37.6%                36
MS Step 21     36.9%                36
MS Step 22     35.8%                36
MS Step 23     34.9%                36
MS Step 24     34.7%                36
MS Step 25     33.9%                36
MS Step 26     33.0%                36
MS Step 27     32.4%                48
MS Step 28     32.1%                48
MS Step 29     31.5%                48
MS Step 30     30.8%                48
MS Step 31     30.1%                48
MS Step 32     29.5%                48
MS Step 33     29.1%                48
MS Step 34     28.6%                48

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of the log weights
(more than 3 or so), and low numbers of unique plans. R-hat values for summary statistics should be between 1 and 1.05.
```
## Additional Notes

## SSD
Performance plot for BVAP:

**<img width="3000" height="1800" alt="bvap_performance" src="https://github.com/user-attachments/assets/46271b7e-d4fc-4e06-8982-e7a9bddb4d56" />**

```
Percentage of Democratic seats by BVAP rank:
   bvap_rank    dem
       <dbl>  <dbl>
 1         1 0     
 2         2 0     
 3         3 0     
 4         4 0     
 5         5 0     
 6         6 0     
 7         7 0     
 8         8 0     
 9         9 0     
10        10 0     
11        11 0     
12        12 0     
13        13 0.0001
14        14 0.0001
15        15 0.0012
16        16 0.0014
17        17 0.0028
18        18 0.0031
19        19 0.0059
20        20 0.0109
21        21 0.0291
22        22 0.0593
23        23 0.132 
24        24 0.223 
25        25 0.328 
26        26 0.405 
27        27 0.462 
28        28 0.500 
29        29 0.502 
30        30 0.590 
31        31 0.819 
32        32 0.956 
33        33 0.996 
34        34 1     
35        35 1     
```
```
Total Black performant districts:
   n_black_perf    n                                                                                                                                                                             
1             3    1
2             4   25
3             5  206
4             6  921
5             7 2629
6             8 3300
7             9 2106
8            10  715
9            11   94
10           12    3

```

## SHD
Performance plot for BVAP:
<img width="3000" height="1800" alt="bvap_performance" src="https://github.com/user-attachments/assets/2be53f4c-6125-401a-b8b4-0e48fffeb298" /> 


```
    bvap_rank    dem
        <dbl>  <dbl>
  1       1   0     
  2       2   0     
  3       3   0     
  4       4   0     
  5       5   0     
  6       6   0     
  7       7   0     
  8       8   0     
  9       9   0     
 10      10   0     
 11      11   0     
 12      11.5 0     
 13      12   0     
 14      13   0     
 15      14   0     
 16      15   0     
 17      16   0     
 18      17   0     
 19      18   0     
 20      19   0     
 21      20   0     
 22      21   0     
 23      22   0     
 24      23   0     
 25      24   0     
 26      25   0     
 27      26   0     
 28      27   0.0001
 29      28   0.0001
 30      29   0.0005
 31      30   0.0003
 32      31   0.0011
 33      32   0.001 
 34      33   0.0013
 35      34   0.0007
 36      35   0.0015
 37      36   0.0015
 38      37   0.0012
 39      38   0.0013
 40      39   0.0025
 41      40   0.0014
 42      41   0.0028
 43      42   0.002 
 44      43   0.0029
 45      44   0.002 
 46      45   0.0029
 47      46   0.0039
 48      47   0.0029
 49      48   0.0029
 50      49   0.0035
 51      50   0.0024
 52      51   0.003 
 53      52   0.0032
 54      53   0.0029
 55      54   0.0041
 56      55   0.005 
 57      56   0.0069
 58      57   0.0106
 59      58   0.012 
 60      59   0.0182
 61      60   0.0281
 62      61   0.0394
 63      62   0.0507
 64      63   0.0658
 65      64   0.0714
 66      65   0.0738
 67      66   0.0788
 68      67   0.0663
 69      68   0.0606
 70      69   0.0654
 71      70   0.0695
 72      71   0.0828
 73      72   0.0768
 74      73   0.0948
 75      74   0.113 
 76      75   0.144 
 77      76   0.182 
 78      77   0.240 
 79      78   0.272 
 80      79   0.284 
 81      80   0.282 
 82      81   0.293 
 83      82   0.340 
 84      83   0.411 
 85      84   0.510 
 86      85   0.627 
 87      86   0.735 
 88      87   0.805 
 89      88   0.834 
 90      89   0.874 
 91      90   0.909 
 92      91   0.948 
 93      92   0.977 
 94      93   0.993 
 95      94   0.998 
 96      95   1.000 
 97      96   1     
 98      97   1     
 99      98   1     
100      99   1     
101     100   1     
102     101   1     
103     102   1     
104     103   1     
105     104   1  
```
```
Total Black performant districts:
   n_black_perf    n                                                                                                                                                                             
1            18   16
2            19   97
3            20  425
4            21 1111
5            22 2059
6            23 2660
7            24 2099
8            25 1105
9            26  344
10           27   74
11           28    9
12           29    1

```

## Checklist

- [x] I have followed the [instructions](https://github.com/alarm-redist/fifty-states/blob/main/CONTRIBUTING.md)
- [x] I have updated the [tracker](https://docs.google.com/spreadsheets/d/1k_tYLoE49W_DCK1tcWbouoYZFI9WD76oayEt5TOmJg4/edit#gid=453387933)
- [x] All `TODO` lines from the template code have been removed
- [x] I have merged in the main branch and then recalculated summary statistics
- [x] I have run `enforce_style()` to format my code
- [x] The documentation copied above is up-to-date 
- [x] There are no data files in this pull request
- [x] None of the file output paths (for the `redist_map` and `redist_plans` objects, and summary statistics) have been edited

@philipwosull

Kara77779 committed:
Add IN-leg-2020

Kara77779 committed:
Merge branch 'main' of https://github.com/alarm-redist/fifty-states into IN-leg-2020

Kara77779 committed:
Add Kentucky legislative redistricting analysis

Kara77779 committed:
Run enforce_style() to format code

Kara77779 committed:
Add Alabama State House (AL_leg_2020) redistricting analysis

Kara77779 committed:
Update doc_AL_leg_2020.md

Kara77779 committed:
Update doc_AL_leg_2020.md

Kara77779 committed:
Update doc_AL_leg_2020.md

Kara77779 committed:
Update doc_KY_leg_2020.md

Kara77779 committed:
Update doc_IN_leg_2020.md

Kara77779 committed:
Update doc_AL_leg_2020.md

ethanjasny commented:
@philipwosull I think we probably want both county constraints to be a bit stronger?

Kara77779 commented:
Hi @christopherkenny @philipwosull 
I ran the full simulation for KY Senate with:

add_constr_total_splits(strength = 3, admin = county)
add_constr_plan_splits(strength = 2, admin = county)
nsims = 2100, runs = 5

The splits metrics look good:

Total county splits: R-hat = 1.018, enacted (~21) within distribution
County splits: enacted (~5) within distribution (~5-10)

However, many other R-hat values are above 1.05, and the summary shows "Chains have not converged."
What would you recommend to improve convergence? Should I:

Increase nsims further?
Adjust mh_accept_per_smc?
Try different constraint strengths?
Something else?

SMC with Merge-Split MCMC Steps: 10,000 sampled plans of 38 districts on 3,693 units
Plans sampled on Linking Edge Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:
• `weight_type` = optimal
• `seq_alpha` = 1
• `pop_temper` = 0
Mergesplit Parameters:
• `total_ms_steps` = 37
• `mh_accept_per_smc` = 33
• `pair_rule` = uniform
Plan diversity 80% range: 0.55 to 0.73
31 rhat values are `NA`
Largest R-hat values for ordered summary statistics:
             adv_16              adv_20              arv_16              arv_20     comp_bbox_reock 
             ❌1.26              ❌1.42             ❌1.433             ❌1.371             ❌1.127 
          comp_edge         comp_polsby       county_splits               e_dem               e_dvs 
            ❌1.056             ❌1.109             ❌1.067             ❌1.384             ❌1.292 
               egap         muni_splits             ndshare                 ndv                 nrv 
            ❌1.363               1.023              ❌1.29             ❌1.372              ❌1.47 
              pbias            plan_dev            pop_aian           pop_asian           pop_black 
            ❌1.141             ❌1.051             ❌1.284             ❌1.481             ❌1.293 
           pop_hisp            pop_nhpi           pop_other         pop_overlap             pop_two 
             ❌1.32             ❌1.391              ❌1.39             ❌1.107             ❌1.343 
          pop_white              pr_dem      pre_16_dem_cli      pre_16_rep_tru      pre_20_dem_bid 
            ❌1.519             ❌1.429             ❌1.252              ❌1.39             ❌1.326 
     pre_20_rep_tru total_county_splits   total_muni_splits           total_vap      uss_16_dem_gra 
            ❌1.345               1.018             ❌1.062             ❌1.243             ❌1.154 
     uss_16_rep_pau      uss_20_dem_mcg      uss_20_rep_mcc            vap_aian           vap_asian 
            ❌1.155             ❌1.331             ❌1.385             ❌1.171             ❌1.535 
          vap_black            vap_hisp            vap_nhpi           vap_other             vap_two 
            ❌1.255             ❌1.293             ❌1.398               ❌1.6             ❌1.432 
          vap_white 
            ❌1.505 
• R-hat ≤ 1.05: 502
• 1.05 < R-hat ≤ 1.1: 550
• R-hat > 1.1: 665
• Total R-hats: 1717
✖ WARNING: Chains have not converged.
Sampling diagnostics for SMC run 1 of 5 (10,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique 
Split 1       469 (22.3%)    100.0%         1.8 1,341 (101%) 
Split 2       211 (10.0%)     98.6%         2.7   779 ( 59%) 
Split 3        201 (9.6%)     96.0%         2.7   437 ( 33%) 
Split 4        208 (9.9%)     94.8%         2.8   443 ( 33%) 
Split 5        196 (9.3%)     92.5%         2.6   442 ( 33%) 
Split 6        193 (9.2%)     88.9%         2.8   446 ( 34%) 
Split 7        173 (8.2%)     86.6%         2.7   450 ( 34%) 
Split 8        202 (9.6%)     84.7%         2.7   486 ( 37%) 
Split 9        180 (8.6%)     81.1%         2.9   483 ( 36%) 
Split 10       208 (9.9%)     80.5%         2.7   469 ( 35%) 
Split 11      249 (11.9%)     77.2%         2.7   557 ( 42%) 
Split 12      248 (11.8%)     74.8%         2.9   556 ( 42%) 
Split 13      323 (15.4%)     72.1%         3.1   558 ( 42%) 
Split 14      325 (15.5%)     71.1%         3.0   597 ( 45%) 
Split 15      387 (18.4%)     68.4%         2.9   625 ( 47%) 
Split 16      424 (20.2%)     63.9%         3.0   683 ( 51%) 
Split 17      462 (22.0%)     62.8%         3.2   684 ( 52%) 
Split 18      507 (24.2%)     60.0%         2.9   687 ( 52%) 
Split 19      448 (21.3%)     58.8%         3.0   737 ( 56%) 
Split 20      501 (23.9%)     56.7%         2.8   722 ( 54%) 
Split 21      466 (22.2%)     55.6%         2.9   750 ( 56%) 
Split 22      483 (23.0%)     55.2%         3.0   723 ( 54%) 
Split 23      440 (21.0%)     54.3%         2.8   730 ( 55%) 
Split 24      469 (22.3%)     50.9%         2.8   723 ( 54%) 
Split 25      480 (22.8%)     48.1%         2.6   751 ( 57%) 
Split 26      449 (21.4%)     46.3%         2.6   770 ( 58%) 
Split 27      472 (22.5%)     45.3%         2.6   779 ( 59%) 
Split 28      479 (22.8%)     44.0%         2.5   769 ( 58%) 
Split 29      514 (24.5%)     43.8%         2.7   774 ( 58%) 
Split 30      564 (26.8%)     41.8%         2.6   803 ( 60%) 
Split 31      616 (29.4%)     41.7%         2.6   810 ( 61%) 
Split 32      640 (30.5%)     37.0%         2.5   857 ( 65%) 
Split 33      644 (30.7%)     35.3%         2.5   875 ( 66%) 
Split 34      736 (35.1%)     31.0%         2.3   913 ( 69%) 
Split 35      861 (41.0%)     29.3%         1.9   979 ( 74%) 
Split 36    1,101 (52.4%)     27.5%         1.7   990 ( 75%) 
Split 37    1,181 (56.3%)     22.9%         1.0 1,019 ( 77%) 
Resample    1,181 (56.3%)       NA%          NA 1,524 (115%) 

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 5 (10,000 samples)
           Acc. rate MS Steps per Plan
MS Step 1      35.5%                33
MS Step 2      24.7%                99
MS Step 3      20.2%               165
MS Step 4      18.5%               165
MS Step 5      17.4%               198
MS Step 6      16.4%               198
MS Step 7      16.1%               231
MS Step 8      15.6%               231
MS Step 9      15.6%               231
MS Step 10     15.2%               231
MS Step 11     15.7%               231
MS Step 12     15.9%               231
MS Step 13     16.4%               231
MS Step 14     17.2%               231
MS Step 15     18.0%               198
MS Step 16     18.6%               198
MS Step 17     19.2%               198
MS Step 18     19.6%               198
MS Step 19     19.8%               198
MS Step 20     19.8%               198
MS Step 21     19.7%               198
MS Step 22     19.8%               198
MS Step 23     19.8%               198
MS Step 24     19.8%               198
MS Step 25     19.9%               198
MS Step 26     19.6%               198
MS Step 27     19.8%               198
MS Step 28     19.9%               198
MS Step 29     19.8%               198
MS Step 30     19.8%               198
MS Step 31     20.1%               198
MS Step 32     20.5%               165
MS Step 33     20.9%               165
MS Step 34     21.0%               165
MS Step 35     20.9%               165
MS Step 36     21.1%               165
MS Step 37     21.4%               165

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of the log
weights (more than 3 or so), and low numbers of unique plans. R-hat values for summary statistics should be
between 1 and 1.05.
• SMC convergence: Increase the number of samples. If you are experiencing low plan diversity or bottlenecks as
well, address those issues first.
<img width="3000" height="5400" alt="validation_20260316_2016" src="https://github.com/user-attachments/assets/e643e31a-5980-48de-9aab-974395632bf2" />

Kara77779 commented:
> I think state senate looks like you're just about there. I would go for the full run with those constraints and slightly oversample. (ie. increase to say 2100 per run) This way we leave a little room to rejection sample at the end, because it seems like this may have a small tail of insufficient performance.

Hi @christopherkenny, sorry for the confusion - I accidentally posted KY Senate results here instead of Alabama. I'll work on the Alabama analysis separately and post the correct results soon.

Kara77779 commented:
> > I think state senate looks like you're just about there. I would go for the full run with those constraints and slightly oversample. (ie. increase to say 2100 per run) This way we leave a little room to rejection sample at the end, because it seems like this may have a small tail of insufficient performance.
> 
> Hi @christopherkenny, sorry for the confusion - I accidentally posted KY Senate results here instead of Alabama. I'll work on the Alabama analysis separately and post the correct results soon.

Hi @christopherkenny @philipwosull
Here are the full simulation results for AL Senate:

nsims = 2100, runs = 5
No additional constraints needed

All R-hat values ≤ 1.05 
Chains converged 
Enacted plan is within the distribution for all key metrics:

County splits 
Total county splits 
Compactness 
Minority VAP share 

SMC with Merge-Split MCMC Steps: 10,000 sampled plans of 35 districts on 1,837 units
Plans sampled on Linking Edge Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:
• `weight_type` = optimal
• `seq_alpha` = 1
• `pop_temper` = 0
Mergesplit Parameters:
• `total_ms_steps` = 34
• `mh_accept_per_smc` = 32
• `pair_rule` = uniform
Plan diversity 80% range: 0.73 to 0.86
23 rhat values are `NA`
Largest R-hat values for ordered summary statistics:
             adv_16              adv_18              adv_20              arv_16              arv_18 
              1.010               1.008               1.010               1.008               1.007 
             arv_20      atg_18_dem_sie      atg_18_rep_mar     comp_bbox_reock           comp_edge 
              1.008               1.007               1.007               1.003               1.002 
        comp_polsby       county_splits               e_dem               e_dvs                egap 
              1.005               1.004               1.001               1.013               1.001 
     gov_18_dem_mad      gov_18_rep_ive         muni_splits             ndshare                 ndv 
              1.011               1.008               1.005               1.013               1.010 
                nrv               pbias            plan_dev            pop_aian           pop_asian 
              1.008               1.003               1.000               1.005               1.005 
          pop_black            pop_hisp            pop_nhpi           pop_other         pop_overlap 
              1.010               1.008               1.008               1.015               1.003 
            pop_two           pop_white              pr_dem      pre_16_dem_cli      pre_16_rep_tru 
              1.007               1.004               1.003               1.011               1.007 
     pre_20_dem_bid      pre_20_rep_tru      sos_18_dem_mil      sos_18_rep_mer total_county_splits 
              1.009               1.007               1.009               1.005               1.003 
  total_muni_splits           total_vap      uss_16_dem_cru      uss_16_rep_she      uss_20_dem_jon 
              1.005               1.001               1.008               1.007               1.010 
     uss_20_rep_tub            vap_aian           vap_asian           vap_black            vap_hisp 
              1.009               1.005               1.006               1.009               1.007 
           vap_nhpi           vap_other             vap_two           vap_white 
              1.007               1.012               1.006               1.004 
• R-hat ≤ 1.05: 1867
• 1.05 < R-hat ≤ 1.1: 0
• R-hat > 1.1: 0
• Total R-hats: 1867
Sampling diagnostics for SMC run 1 of 5 (10,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique 
Split 1       981 (46.7%)    100.0%        1.25 1,336 (101%) 
Split 2     1,083 (51.6%)     97.6%        1.21 1,075 ( 81%) 
Split 3     1,185 (56.4%)     94.4%        1.12 1,082 ( 82%) 
Split 4     1,222 (58.2%)     92.1%        1.03 1,102 ( 83%) 
Split 5     1,368 (65.2%)     88.5%        0.93 1,114 ( 84%) 
Split 6     1,333 (63.5%)     85.3%        0.88 1,131 ( 85%) 
Split 7     1,443 (68.7%)     81.7%        0.80 1,133 ( 85%) 
Split 8     1,456 (69.3%)     81.3%        0.78 1,178 ( 89%) 
Split 9     1,511 (71.9%)     76.6%        0.71 1,168 ( 88%) 
Split 10    1,578 (75.2%)     74.0%        0.67 1,167 ( 88%) 
Split 11    1,589 (75.7%)     72.3%        0.65 1,232 ( 93%) 
Split 12    1,599 (76.1%)     69.0%        0.62 1,197 ( 90%) 
Split 13    1,648 (78.5%)     66.3%        0.59 1,223 ( 92%) 
Split 14    1,666 (79.3%)     60.7%        0.57 1,235 ( 93%) 
Split 15    1,702 (81.1%)     61.9%        0.54 1,237 ( 93%) 
Split 16    1,703 (81.1%)     58.9%        0.52 1,253 ( 94%) 
Split 17    1,735 (82.6%)     56.5%        0.50 1,210 ( 91%) 
Split 18    1,780 (84.8%)     52.1%        0.47 1,246 ( 94%) 
Split 19    1,776 (84.6%)     52.5%        0.47 1,256 ( 95%) 
Split 20    1,800 (85.7%)     50.5%        0.45 1,277 ( 96%) 
Split 21    1,813 (86.3%)     48.4%        0.43 1,264 ( 95%) 
Split 22    1,834 (87.3%)     45.6%        0.41 1,232 ( 93%) 
Split 23    1,860 (88.6%)     42.8%        0.39 1,241 ( 93%) 
Split 24    1,864 (88.8%)     44.3%        0.39 1,248 ( 94%) 
Split 25    1,861 (88.6%)     41.2%        0.39 1,248 ( 94%) 
Split 26    1,889 (90.0%)     38.2%        0.36 1,237 ( 93%) 
Split 27    1,906 (90.8%)     37.7%        0.35 1,275 ( 96%) 
Split 28    1,906 (90.8%)     35.7%        0.35 1,242 ( 94%) 
Split 29    1,931 (92.0%)     34.2%        0.32 1,247 ( 94%) 
Split 30    1,943 (92.5%)     34.4%        0.31 1,244 ( 94%) 
Split 31    1,958 (93.2%)     32.7%        0.29 1,230 ( 93%) 
Split 32    1,967 (93.7%)     32.2%        0.28 1,232 ( 93%) 
Split 33    1,991 (94.8%)     29.3%        0.26 1,197 ( 90%) 
Split 34    2,000 (95.2%)     29.2%        0.25 1,185 ( 89%) 
Resample    2,000 (95.2%)       NA%          NA 1,916 (144%) 

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 5 (10,000 samples)
           Acc. rate MS Steps per Plan
MS Step 1      52.0%                32
MS Step 2      50.7%                64
MS Step 3      49.6%                64
MS Step 4      48.9%                96
MS Step 5      48.2%                96
MS Step 6      47.6%                96
MS Step 7      47.1%                96
MS Step 8      46.3%                96
MS Step 9      45.1%                96
MS Step 10     44.8%                96
MS Step 11     43.9%                96
MS Step 12     42.8%                96
MS Step 13     42.3%                96
MS Step 14     41.3%                96
MS Step 15     41.0%                96
MS Step 16     40.1%                96
MS Step 17     39.5%                96
MS Step 18     38.8%                96
MS Step 19     37.9%                96
MS Step 20     37.3%                96
MS Step 21     36.5%                96
MS Step 22     35.7%                96
MS Step 23     35.0%                96
MS Step 24     34.4%                96
MS Step 25     33.9%                96
MS Step 26     33.1%                96
MS Step 27     32.7%               128
MS Step 28     31.9%               128
MS Step 29     31.5%               128
MS Step 30     31.0%               128
MS Step 31     30.3%               128
MS Step 32     29.5%               128
MS Step 33     29.3%               128
MS Step 34     28.8%               128

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of the log
weights (more than 3 or so), and low numbers of unique plans. R-hat values for summary statistics should be
between 1 and 1.05.

<img width="3000" height="5400" alt="validation_20260318_2203" src="https://github.com/user-attachments/assets/01b2d995-b984-4fea-beaf-3ee39cfc5695" />
